File: experiments/pixel-gear/gel.ts
import type { PixelGearState } from "./gear.ts";
import type { PixelGearKernel } from "./kernel.ts";

export interface PixelGelCluster {
  readonly id: number;
  /** Local cell centers stored as x/y pairs. Rebuilt only when the cluster fractures. */
  readonly cells: Float32Array;
  /** SoA boundary coordinates consumed directly by the Zig SIMD loop. */
  readonly boundaryX: Float32Array;
  readonly boundaryY: Float32Array;
  centerX: number;
  centerY: number;
  angle: number;
  velocityX: number;
  velocityY: number;
  angularVelocity: number;
  stress: number;
  strength: number;
  fractureCooldown: number;
}

export interface PixelGelStepOptions {
  readonly gravity?: number;
  readonly damping?: number;
  readonly stressDecay?: number;
  readonly minimumFragmentCells?: number;
}

export interface PixelGelStepResult {
  readonly clusters: PixelGelCluster[];
  readonly boundaryChecks: number;
  readonly contacts: number;
  readonly fractures: number;
}

let nextClusterId = 1;

export function createPixelGelBlock(
  centerX: number,
  centerY: number,
  width: number,
  height: number,
  options: { readonly strength?: number } = {},
): PixelGelCluster {
  if (
    !Number.isSafeInteger(width) || !Number.isSafeInteger(height) || width <= 0 || height <= 0 ||
    !Number.isFinite(centerX) || !Number.isFinite(centerY)
  ) throw new RangeError("pixel gel block parameters are invalid");
  const cells = new Float32Array(width * height * 2);
  let cursor = 0;
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      cells[cursor++] = x - (width - 1) / 2;
      cells[cursor++] = y - (height - 1) / 2;
    }
  }
  return createCluster(cells, centerX, centerY, options.strength ?? 18);
}

export function createPixelGelBlob(
  centerX: number,
  centerY: number,
  radiusX: number,
  radiusY: number,
  options: { readonly strength?: number } = {},
): PixelGelCluster {
  if (
    !Number.isFinite(centerX) || !Number.isFinite(centerY) ||
    !Number.isFinite(radiusX) || !Number.isFinite(radiusY) || radiusX < 1 || radiusY < 1
  ) throw new RangeError("pixel gel blob parameters are invalid");
  const points: number[] = [];
  for (let y = -Math.ceil(radiusY); y <= Math.ceil(radiusY); y++) {
    for (let x = -Math.ceil(radiusX); x <= Math.ceil(radiusX); x++) {
      if ((x / radiusX) ** 2 + (y / radiusY) ** 2 <= 1) points.push(x, y);
    }
  }
  return createCluster(new Float32Array(points), centerX, centerY, options.strength ?? 18);
}

/**
 * Advances bonded gel aggregates. Normal motion only checks cached boundary
 * cells; the complete cell set is visited only by the fracture path.
 */
export function stepPixelGel(
  kernel: PixelGearKernel,
  clusters: readonly PixelGelCluster[],
  width: number,
  height: number,
  gear: PixelGearState | null,
  options: PixelGelStepOptions = {},
): PixelGelStepResult {
  if (!Number.isSafeInteger(width) || !Number.isSafeInteger(height) || width <= 0 || height <= 0) {
    throw new RangeError("pixel gel world dimensions must be positive safe integers");
  }
  const gravity = finiteOption(options.gravity, 0.075, "gravity");
  const damping = finiteOption(options.damping, 0.992, "damping");
  const stressDecay = finiteOption(options.stressDecay, 0.965, "stress decay");
  const minimumFragmentCells = options.minimumFragmentCells ?? 18;
  if (!Number.isSafeInteger(minimumFragmentCells) || minimumFragmentCells < 1) {
    throw new RangeError("minimum gel fragment size must be a positive safe integer");
  }

  const next: PixelGelCluster[] = [];
  let boundaryChecks = 0;
  let contacts = 0;
  let fractures = 0;
  for (const cluster of clusters) {
    cluster.velocityY += gravity;
    cluster.centerX += cluster.velocityX;
    cluster.centerY += cluster.velocityY;
    cluster.angle += cluster.angularVelocity;
    cluster.velocityX *= damping;
    cluster.velocityY *= damping;
    cluster.angularVelocity *= damping;
    cluster.stress *= stressDecay;
    if (cluster.fractureCooldown > 0) cluster.fractureCooldown--;

    const cosine = Math.cos(cluster.angle);
    const sine = Math.sin(cluster.angle);
    const scan = kernel.scanGelBoundary(
      cluster.boundaryX,
      cluster.boundaryY,
      { centerX: cluster.centerX, centerY: cluster.centerY, cosine, sine },
      gear,
    );
    boundaryChecks += scan.checks;
    contacts += scan.contacts;
    cluster.stress += scan.stressDelta;
    const clusterContacts = scan.contacts;
    const correctionX = scan.correctionX;
    const correctionY = scan.correctionY;
    let contactNormalX = scan.contactNormalX;
    let contactNormalY = scan.contactNormalY;
    let contactLocalX = scan.contactLocalX;
    let contactLocalY = scan.contactLocalY;
    const minimumX = scan.minimumX;
    const maximumX = scan.maximumX;
    const maximumY = scan.maximumY;

    if (clusterContacts > 0 && gear !== null) {
      const inverseContacts = 1 / clusterContacts;
      cluster.centerX += correctionX * inverseContacts;
      cluster.centerY += correctionY * inverseContacts;
      const normalLength = Math.max(0.001, Math.hypot(contactNormalX, contactNormalY));
      contactNormalX /= normalLength;
      contactNormalY /= normalLength;
      contactLocalX *= inverseContacts;
      contactLocalY *= inverseContacts;
      const rotation = Math.sign(gear.angularVelocity) || 1;
      const tangentX = -contactNormalY * rotation;
      const tangentY = contactNormalX * rotation;
      const drive = Math.abs(gear.angularVelocity) * (gear.radius + gear.toothDepth);
      cluster.velocityX += tangentX * drive * 0.025 + contactNormalX * 0.035;
      cluster.velocityY += tangentY * drive * 0.025 + contactNormalY * 0.035;
      cluster.angularVelocity += gear.angularVelocity * 0.018;
    }

    if (minimumX < 1) {
      cluster.centerX += 1 - minimumX;
      cluster.velocityX = Math.abs(cluster.velocityX) * 0.12;
    } else if (maximumX > width - 2) {
      cluster.centerX -= maximumX - (width - 2);
      cluster.velocityX = -Math.abs(cluster.velocityX) * 0.12;
    }
    if (maximumY > height - 2) {
      cluster.centerY -= maximumY - (height - 2);
      cluster.velocityY = -Math.abs(cluster.velocityY) * 0.08;
      cluster.velocityX *= 0.82;
      cluster.angularVelocity *= 0.82;
    }

    if (
      clusterContacts > 0 && cluster.stress > cluster.strength &&
      cluster.fractureCooldown === 0 &&
      cluster.cells.length / 2 >= minimumFragmentCells * 2
    ) {
      const fragments = fractureCluster(
        kernel,
        cluster,
        contactNormalX,
        contactNormalY,
        contactLocalX,
        contactLocalY,
        minimumFragmentCells,
      );
      if (fragments !== null) {
        next.push(...fragments);
        fractures++;
        continue;
      }
    }
    next.push(cluster);
  }
  return { clusters: next, boundaryChecks, contacts, fractures };
}

function fractureCluster(
  kernel: PixelGearKernel,
  cluster: PixelGelCluster,
  worldNormalX: number,
  worldNormalY: number,
  contactLocalX: number,
  contactLocalY: number,
  minimumFragmentCells: number,
): PixelGelCluster[] | null {
  const cosine = Math.cos(cluster.angle);
  const sine = Math.sin(cluster.angle);
  const localNormalX = worldNormalX * cosine + worldNormalY * sine;
  const localNormalY = -worldNormalX * sine + worldNormalY * cosine;
  const tangentX = -localNormalY;
  const tangentY = localNormalX;
  const contactProjection = contactLocalX * localNormalX + contactLocalY * localNormalY;
  const maximumOffset = Math.sqrt(cluster.cells.length / 2) * 0.22;
  const crackOffset = clamp(contactProjection * 0.28, -maximumOffset, maximumOffset);
  const classification = kernel.classifyGelFracture(cluster.cells, {
    normalX: localNormalX,
    normalY: localNormalY,
    tangentX,
    tangentY,
    offset: crackOffset,
    clusterId: cluster.id,
  });
  const cellCount = cluster.cells.length / 2;
  if (
    classification.negativeCount < minimumFragmentCells ||
    cellCount - classification.negativeCount < minimumFragmentCells
  ) return null;
  const negative: number[] = [];
  const positive: number[] = [];
  for (let index = 0; index < cluster.cells.length; index += 2) {
    const x = cluster.cells[index]!;
    const y = cluster.cells[index + 1]!;
    const target = classification.sides[index / 2] === 1 ? negative : positive;
    target.push(x, y);
  }
  const [negativeComponents, positiveComponents] = normalizeCrackComponents(negative, positive);
  const fragments: PixelGelCluster[] = [];
  for (const component of negativeComponents) {
    fragments.push(createFragment(cluster, component, -1, localNormalX, localNormalY));
  }
  for (const component of positiveComponents) {
    fragments.push(createFragment(cluster, component, 1, localNormalX, localNormalY));
  }
  return fragments.length < 2 ? null : fragments;
}

function normalizeCrackComponents(
  negative: readonly number[],
  positive: readonly number[],
): readonly [Float32Array[], Float32Array[]] {
  const negativeComponents = collectConnectedComponents(negative);
  const positiveComponents = collectConnectedComponents(positive);
  const minimumChipCells = 4;
  const negativeIslands = negativeComponents.filter((part) => part.length / 2 < minimumChipCells);
  const positiveIslands = positiveComponents.filter((part) => part.length / 2 < minimumChipCells);
  return [
    collectConnectedComponents(flattenComponents(
      negativeComponents.filter((part) => part.length / 2 >= minimumChipCells),
      positiveIslands,
    )),
    collectConnectedComponents(flattenComponents(
      positiveComponents.filter((part) => part.length / 2 >= minimumChipCells),
      negativeIslands,
    )),
  ];
}

function flattenComponents(
  main: readonly Float32Array[],
  additions: readonly Float32Array[],
): number[] {
  const points: number[] = [];
  for (const component of [...main, ...additions]) {
    for (const value of component) points.push(value);
  }
  return points;
}

function collectConnectedComponents(points: readonly number[]): Float32Array[] {
  const remaining = new Map<string, readonly [number, number]>();
  for (let index = 0; index < points.length; index += 2) {
    const point = [points[index]!, points[index + 1]!] as const;
    remaining.set(cellKey(point[0], point[1]), point);
  }
  const components: Float32Array[] = [];
  while (remaining.size > 0) {
    const first = remaining.values().next().value as readonly [number, number];
    const pending = [first];
    const component: number[] = [];
    remaining.delete(cellKey(first[0], first[1]));
    while (pending.length > 0) {
      const point = pending.pop()!;
      component.push(point[0], point[1]);
      for (
        const [x, y] of [
          [point[0] - 1, point[1]],
          [point[0] + 1, point[1]],
          [point[0], point[1] - 1],
          [point[0], point[1] + 1],
        ] as const
      ) {
        const neighbor = remaining.get(cellKey(x, y));
        if (neighbor === undefined) continue;
        remaining.delete(cellKey(x, y));
        pending.push(neighbor);
      }
    }
    components.push(new Float32Array(component));
  }
  return components;
}

function createFragment(
  parent: PixelGelCluster,
  parentLocalCells: Float32Array,
  side: number,
  localNormalX: number,
  localNormalY: number,
): PixelGelCluster {
  let centerLocalX = 0;
  let centerLocalY = 0;
  for (let index = 0; index < parentLocalCells.length; index += 2) {
    centerLocalX += parentLocalCells[index]!;
    centerLocalY += parentLocalCells[index + 1]!;
  }
  const count = parentLocalCells.length / 2;
  centerLocalX /= count;
  centerLocalY /= count;
  const cells = new Float32Array(parentLocalCells.length);
  for (let index = 0; index < parentLocalCells.length; index += 2) {
    cells[index] = parentLocalCells[index]! - centerLocalX;
    cells[index + 1] = parentLocalCells[index + 1]! - centerLocalY;
  }
  const cosine = Math.cos(parent.angle);
  const sine = Math.sin(parent.angle);
  const offsetX = centerLocalX * cosine - centerLocalY * sine;
  const offsetY = centerLocalX * sine + centerLocalY * cosine;
  const fragment = createCluster(
    cells,
    parent.centerX + offsetX,
    parent.centerY + offsetY,
    parent.strength * 2.2,
  );
  fragment.angle = parent.angle;
  fragment.velocityX = parent.velocityX - parent.angularVelocity * offsetY;
  fragment.velocityY = parent.velocityY + parent.angularVelocity * offsetX;
  fragment.angularVelocity = parent.angularVelocity + side * 0.012;
  fragment.fractureCooldown = 24;
  const worldNormalForFragmentX = localNormalX * cosine - localNormalY * sine;
  const worldNormalForFragmentY = localNormalX * sine + localNormalY * cosine;
  fragment.centerX += worldNormalForFragmentX * side * 1.35;
  fragment.centerY += worldNormalForFragmentY * side * 1.35;
  return fragment;
}

function createCluster(
  cells: Float32Array,
  centerX: number,
  centerY: number,
  strength: number,
): PixelGelCluster {
  if (!Number.isFinite(strength) || strength <= 0) {
    throw new RangeError("pixel gel strength must be positive");
  }
  const [boundaryX, boundaryY] = collectBoundary(cells);
  return {
    id: nextClusterId++,
    cells,
    boundaryX,
    boundaryY,
    centerX,
    centerY,
    angle: 0,
    velocityX: 0,
    velocityY: 0,
    angularVelocity: 0,
    stress: 0,
    strength,
    fractureCooldown: 0,
  };
}

function collectBoundary(cells: Float32Array): readonly [Float32Array, Float32Array] {
  const occupied = new Set<string>();
  for (let index = 0; index < cells.length; index += 2) {
    occupied.add(cellKey(cells[index]!, cells[index + 1]!));
  }
  const boundaryX: number[] = [];
  const boundaryY: number[] = [];
  for (let index = 0; index < cells.length; index += 2) {
    const x = cells[index]!;
    const y = cells[index + 1]!;
    if (
      !occupied.has(cellKey(x - 1, y)) || !occupied.has(cellKey(x + 1, y)) ||
      !occupied.has(cellKey(x, y - 1)) || !occupied.has(cellKey(x, y + 1))
    ) {
      boundaryX.push(x);
      boundaryY.push(y);
    }
  }
  return [new Float32Array(boundaryX), new Float32Array(boundaryY)];
}

function cellKey(x: number, y: number): string {
  // Fragments are recentered in Float32 storage. Quantization removes the
  // sub-ULP drift while preserving the unit-spaced bonded lattice.
  return `${Math.round(x * 1_024)}:${Math.round(y * 1_024)}`;
}

function clamp(value: number, minimum: number, maximum: number): number {
  return Math.max(minimum, Math.min(maximum, value));
}

function finiteOption(value: number | undefined, fallback: number, name: string): number {
  const resolved = value ?? fallback;
  if (!Number.isFinite(resolved)) throw new RangeError(`pixel gel ${name} must be finite`);
  return resolved;
}

File: experiments/pixel-gear/zig/gel.zig
const std = @import("std");
const gear_module = @import("gear.zig");

const F32x4 = @Vector(4, f32);
const Gear = gear_module.Gear;

pub const BoundaryTransform = struct {
    center_x: f32,
    center_y: f32,
    cosine: f32,
    sine: f32,
};

pub const FracturePlane = struct {
    normal_x: f32,
    normal_y: f32,
    tangent_x: f32,
    tangent_y: f32,
    offset: f32,
    cluster_id: u32,
};

pub const BoundaryResult = struct {
    checks: u32 = 0,
    contacts: u32 = 0,
    minimum_x: f32 = std.math.inf(f32),
    maximum_x: f32 = -std.math.inf(f32),
    maximum_y: f32 = -std.math.inf(f32),
    correction_x: f32 = 0,
    correction_y: f32 = 0,
    contact_normal_x: f32 = 0,
    contact_normal_y: f32 = -1,
    contact_local_x: f32 = 0,
    contact_local_y: f32 = 0,
    stress_delta: f32 = 0,
};

fn includeBounds(result: *BoundaryResult, world_x: f32, world_y: f32) void {
    result.minimum_x = @min(result.minimum_x, world_x);
    result.maximum_x = @max(result.maximum_x, world_x);
    result.maximum_y = @max(result.maximum_y, world_y);
}

fn includeContact(
    result: *BoundaryResult,
    gear: Gear,
    world_x: f32,
    world_y: f32,
    local_x: f32,
    local_y: f32,
) void {
    const cell_x: i32 = @intFromFloat(@floor(world_x));
    const cell_y: i32 = @intFromFloat(@floor(world_y));
    if (!gear_module.contains(gear, cell_x, cell_y)) return;

    const delta_x = world_x - gear.center_x;
    const delta_y = world_y - gear.center_y;
    const distance = @max(@as(f32, 0.001), @sqrt(delta_x * delta_x + delta_y * delta_y));
    const normal_x = delta_x / distance;
    const normal_y = delta_y / distance;
    const rotation: f32 = if (gear.angular_velocity < 0) -1 else 1;
    const tangent_x = -normal_y * rotation;
    const tangent_y = normal_x * rotation;
    const surface_speed = @abs(gear.angular_velocity) * distance;

    result.correction_x += normal_x * 0.85 + tangent_x * surface_speed * 0.08;
    result.correction_y += normal_y * 0.85 + tangent_y * surface_speed * 0.08;
    result.contact_normal_x += normal_x;
    result.contact_normal_y += normal_y;
    result.contact_local_x += local_x;
    result.contact_local_y += local_y;
    result.stress_delta += 0.12 + surface_speed * 0.18;
    result.contacts += 1;
}

pub fn scanBoundary(
    local_x: []const f32,
    local_y: []const f32,
    transform: BoundaryTransform,
    gear: ?Gear,
) BoundaryResult {
    std.debug.assert(local_x.len == local_y.len);
    var result = BoundaryResult{ .checks = @intCast(local_x.len) };
    const center_x: F32x4 = @splat(transform.center_x);
    const center_y: F32x4 = @splat(transform.center_y);
    const cosine: F32x4 = @splat(transform.cosine);
    const sine: F32x4 = @splat(transform.sine);
    var index: usize = 0;

    while (index + 4 <= local_x.len) : (index += 4) {
        const xs: *align(1) const F32x4 = @ptrCast(local_x.ptr + index);
        const ys: *align(1) const F32x4 = @ptrCast(local_y.ptr + index);
        const world_x = center_x + xs.* * cosine - ys.* * sine;
        const world_y = center_y + xs.* * sine + ys.* * cosine;

        inline for (0..4) |lane| includeBounds(&result, world_x[lane], world_y[lane]);
        if (gear) |active_gear| {
            const half: F32x4 = @splat(0.5);
            const gear_x: F32x4 = @splat(active_gear.center_x);
            const gear_y: F32x4 = @splat(active_gear.center_y);
            const cell_x = @floor(world_x) + half;
            const cell_y = @floor(world_y) + half;
            const delta_x = cell_x - gear_x;
            const delta_y = cell_y - gear_y;
            const reach: F32x4 = @splat(active_gear.radius + active_gear.tooth_depth);
            const candidates = delta_x * delta_x + delta_y * delta_y <= reach * reach;
            const mask: u4 = @bitCast(candidates);
            inline for (0..4) |lane| {
                if (mask & (@as(u4, 1) << lane) != 0) {
                    includeContact(
                        &result,
                        active_gear,
                        world_x[lane],
                        world_y[lane],
                        xs.*[lane],
                        ys.*[lane],
                    );
                }
            }
        }
    }

    while (index < local_x.len) : (index += 1) {
        const x = local_x[index];
        const y = local_y[index];
        const world_x = transform.center_x + x * transform.cosine - y * transform.sine;
        const world_y = transform.center_y + x * transform.sine + y * transform.cosine;
        includeBounds(&result, world_x, world_y);
        if (gear) |active_gear| includeContact(&result, active_gear, world_x, world_y, x, y);
    }
    return result;
}

fn crackNoise(tangent_position: f32, cluster_id: u32) f32 {
    const seed = @as(f32, @floatFromInt(cluster_id)) * 0.731;
    return @sin(tangent_position * 0.32 + seed) * 1.55 +
        @sin(tangent_position * 0.13 - seed * 0.7) * 0.8;
}

pub fn classifyFracture(
    local_x: []const f32,
    local_y: []const f32,
    sides: []u8,
    plane: FracturePlane,
) u32 {
    std.debug.assert(local_x.len == local_y.len and local_x.len == sides.len);
    const normal_x: F32x4 = @splat(plane.normal_x);
    const normal_y: F32x4 = @splat(plane.normal_y);
    const tangent_x: F32x4 = @splat(plane.tangent_x);
    const tangent_y: F32x4 = @splat(plane.tangent_y);
    const offset: F32x4 = @splat(plane.offset);
    var negative_count: u32 = 0;
    var index: usize = 0;

    while (index + 4 <= local_x.len) : (index += 4) {
        const xs: *align(1) const F32x4 = @ptrCast(local_x.ptr + index);
        const ys: *align(1) const F32x4 = @ptrCast(local_y.ptr + index);
        const tangent_position = xs.* * tangent_x + ys.* * tangent_y;
        const signed_distance = xs.* * normal_x + ys.* * normal_y - offset;
        inline for (0..4) |lane| {
            const negative = signed_distance[lane] +
                crackNoise(tangent_position[lane], plane.cluster_id) < 0;
            sides[index + lane] = @intFromBool(negative);
            negative_count += @intFromBool(negative);
        }
    }

    while (index < local_x.len) : (index += 1) {
        const tangent_position = local_x[index] * plane.tangent_x + local_y[index] * plane.tangent_y;
        const signed_distance = local_x[index] * plane.normal_x + local_y[index] * plane.normal_y -
            plane.offset + crackNoise(tangent_position, plane.cluster_id);
        const negative = signed_distance < 0;
        sides[index] = @intFromBool(negative);
        negative_count += @intFromBool(negative);
    }
    return negative_count;
}

test "boundary scan transforms four lanes and reports world bounds" {
    const local_x = [_]f32{ -2, -1, 1, 2 };
    const local_y = [_]f32{ 1, -3, 4, 0 };
    const result = scanBoundary(
        &local_x,
        &local_y,
        .{ .center_x = 10, .center_y = 20, .cosine = 1, .sine = 0 },
        null,
    );

    try std.testing.expectEqual(@as(u32, 4), result.checks);
    try std.testing.expectEqual(@as(u32, 0), result.contacts);
    try std.testing.expectApproxEqAbs(@as(f32, 8), result.minimum_x, 0.0001);
    try std.testing.expectApproxEqAbs(@as(f32, 12), result.maximum_x, 0.0001);
    try std.testing.expectApproxEqAbs(@as(f32, 24), result.maximum_y, 0.0001);
}

test "fracture classification writes a side byte for vector lanes and tail" {
    const local_x = [_]f32{ -100, -50, 50, 100, 120 };
    const local_y = [_]f32{ 0, 0, 0, 0, 0 };
    var sides = [_]u8{0} ** local_x.len;
    const negative_count = classifyFracture(
        &local_x,
        &local_y,
        &sides,
        .{
            .normal_x = 1,
            .normal_y = 0,
            .tangent_x = 0,
            .tangent_y = 1,
            .offset = 0,
            .cluster_id = 1,
        },
    );

    try std.testing.expectEqual(@as(u32, 2), negative_count);
    try std.testing.expectEqualSlices(u8, &[_]u8{ 1, 1, 0, 0, 0 }, &sides);
}
